File: Algorithms/77. Combinations/solution.go
package problem77

func combine(n int, k int) [][]int {
	combinations := [][]int{}
	helper(n, k, 1, []int{}, &combinations)
	return combinations
}

func helper(n int, k int, s int, cur []int, res *[][]int) {
	if len(cur) == k {
		c := make([]int, len(cur))
		copy(c, cur)
		*res = append(*res, c)
		return
	}
	for i := s; i <= n; i++ {
		helper(n, k, i+1, append(cur, i), res)
	}
}

File: Algorithms/77. Combinations/solution.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function (n, k) {
  const result = [];
  helper(n, k, 1, [], result);
  return result;
};

function helper(n, k, s, cur, res) {
  if (cur.length === k) return res.push(cur.slice());
  for (let i = s; i <= n; i++) {
    cur.push(i);
    helper(n, k, i + 1, cur, res);
    cur.pop();
  }
}
